smartsim/experiment.py

```diff
@@ -218,15 +223,73 @@ def execute_dispatch(job: Job) -> LaunchedJobID:
                 launch_config = dispatch.create_new_launcher_configuration(
                     for_experiment=self, with_arguments=args
                 )
-            # Save the underlying launcher instance. That way we do not need to
-            # spin up a launcher instance for each individual job, and it makes
-            # it easier to monitor job statuses
+            id_ = launch_config.start(exe, env)
+            # Save the underlying launcher instance and launched job id. That
+            # way we do not need to spin up a launcher instance for each
+            # individual job, and the experiment can monitor job statuses.
             # pylint: disable-next=protected-access
-            self._active_launchers.add(launch_config._adapted_launcher)
-            return launch_config.start(exe, env)
+            self._active_launchers[launch_config._adapted_launcher].add(id_)
+            return id_
 
         return execute_dispatch(job), *map(execute_dispatch, jobs)
 
+    def get_status(self, *ids: LaunchedJobID) -> tuple[SmartSimStatus, ...]:
```

Comment:
This function is good, but I wonder how performant it will be if it is called very often with a large amount of jobs present.  

Would it be easier / more efficient if the initial mapping was job_id -> [launcher]?  Or maybe it is better if we put the data structures and logic for mananging this mapping in a class so implmeentation details are hidden a bit.  Maybe something like ``class JobInventory`` (better class names welcome)

Reply:
Moved to a new `LaunchHistory` data structure that stores a mapping of `job_id` -> `launcher` that issued the id